src/awkward/operations/ak_cartesian.py

```diff
@@ -398,10 +402,14 @@ def apply(x, i):
 
         def getfunction3(inputs, depth, **kwargs):
             if depth == posaxis + len(new_arrays):
-                if all(len(x) == 0 for x in inputs):
+                if all(
+                    x.length is not unknown_length and x.length == 0 for x in inputs
+                ):
                     inputs = [
                         x.content
-                        if isinstance(x, ak.contents.RegularArray) and x.size == 1
+                        if isinstance(x, ak.contents.RegularArray)
+                        and x.size is not unknown_length
+                        and x.size == 1
                         else x
```

Comment:
@jpivarski / @ioanaif do you remember what the purpose of this step is? I can spend some time figuring it out, but if you can think off the top of your head, it would be helpful!

Reply:
Note to self: make this blow up to find a reproducer & counterexample. See what typetracer does thereafter.